rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+
+### `router` / `router-api` are complicated and hard to maintain
+[router] routes requests for pages on the site to the correct application. `router` is complicated and hard to maintain. Moving towards a single application will remove the need for `router` entirely.
+
+`router` is a Go application that copies a database of URLs on GOV.UK from another application called [router-api]. `router` uses that database to route requests to the correct rendering application.
+
+Go is not a language that many developers on GOV.UK are comfortable with, which makes updating the application harder.
+
+The relationship between [router and router-api] is quite tangled. `router-api` is a Ruby on Rails application. The database that it shares with router is a Mongo database.  The version of Mongo we’re using is an old one, and has had to be pinned to keep [compatibility with the Mongo driver].
+
+`content-store` registers routes in `router-api`. This means that the same content data is stored in multiple places and has the potential to get out of sync.
+
+There was an attempt to switch `router-api` to [use PostgreSQL] but this was paused indefinitely due to other priorities and the complexity of the work.
+
+### Local preview and sharing changes to GOV.UK is difficult
+Manually testing user journeys from a page is difficult, particularly where code changes are spread across more than one application. To test locally, developers have to have multiple applications running which slows down their local machines.
+
+Having fewer apps would also mean that GOV.UK Docker, our system for running apps locally, could be simplified or potentially removed entirely.
+
+Sharing changes for non-engineering disciplines to preview is also difficult. This can be done by deploying branches to integration, but these branches are frequently overwritten by dependency updates.
+
+We can use Heroku for previews but this is also complicated by multiple apps. Previewing a user journey that involves pages from more than one app cannot be done in Heroku because links from a page in one app do not link to another. For example, a journey from the homepage to a search page and then a mainstream browse page crosses multiple app boundaries, and the links would 404 in Heroku. Previews can be done using Heroku, but this would be greatly simplified by having only a single application.
+
+### `static` and `slimmer` are complicated and inflexible
+GOV.UK page layouts are rendered by `static` using the [slimmer] gem. This includes the header and footer common to all pages, even though the code for those elements comes from [govuk_publishing_components]. Having a single application would allow us to remove our dependency upon `static`.
+
+`static` and `slimmer` are both quite complicated and not well understood among the current developer community. The details are beyond the scope of this document, but several specific problems can be identified.
```

Comment:
Eventually being able to remove `static` and `slimmer` is definitely one of the points of this RFC and a goal of these changes. 

We're in a bit of a strange situation where the layouts live in `govuk_publishing_components`, but are called via `static`. In the case of the [standard layout](https://github.com/alphagov/static/blob/03878acd740d0407fb35c807d5c53f7dccebf320/app/views/root/_gem_base.html.erb), the only extra work that's happening is rendering the global and emergency banners.

The current setup of `static` and `slimmer` where the templates are passed about as string, also make it very difficult to have dynamic headers. This is something that could cause a bit of pain when we need to integrate with One Login.

But the details of how static and slimmer are removed or replaced should have its own discussion.

Reply:
It does seem like that is many years off to combine the apps to a point where you could kill static, I think it'd be much more pragmatic to investigate killing static off without needing to merge all the apps if that is a bigger prize than just having less apps. Having to consolidate all the apps as a pre-requisite to not needing static does seem a bit of nuclear option. 

Like you say, it doesn't really seem to do a lot anymore (I hadn't realised how little, I thought it still had links configured in it, but they seem ). It seems that the only benefit of having apps use that over just rendering the layout themselves is that it means we only have to coordinate statics update of govuk_publishing_components over having to coordinate updating each frontend app when header or footer change (as an aside I do think it's quite an architecture smell the degree of content in govuk_publishing_components, really think we could benefit from apps able to inject that in from a content item).

I would be inclined to suggest that as part of the work to try make frontend the "main" frontend app, we try just dropping static from it and seeing how things go with using govuk_publishing_components directly and see if we can avoid waiting for the long migration.

Reply:
Removing static is a problem that has been investigated before: https://github.com/alphagov/govuk-rfcs/pull/84
It comes a bit unstuck when it comes to the banners. If static is replaced with a gem, then you have the issue of having to do a gem update and deploy in 8 rendering apps. That's better now that we have auto-merging and deployment on most of the apps, but could still leave things in an inconsistent state if a deployment fails.

There also the issue of duplicating assets if static is removed now before doing any consolidation.

I think that consolidating more apps gets us closer to the place where decommissioning static starts looking possible.

Reply:
Yeah there's further work after that RFC too:

https://github.com/alphagov/govuk-rfcs/pull/118 - which explored that content and other global aspects aren't hardcoded but are part of data from Content Store, which would allow banners to go out without being hardcoded.

And for assets there is the unimplemented approach to use a shared folder from a couple of RFCs: https://github.com/alphagov/govuk-rfcs/blob/109d05df9dcb9f06945ffa512f65d185b05f45a4/rfc-149-switch-to-per-page-asset-loading.md#make-assets-load-from-a-shared-folder and https://github.com/alphagov/govuk-rfcs/blob/109d05df9dcb9f06945ffa512f65d185b05f45a4/rfc-091-sharing-assets.md

Reply:
Being better in a better position to remove static and slimmer is one of the justifications of doing this work. The less rendering apps we have to deal with the more options we have. I don't we can omit mention of them entirely.